[BC break] Refactors JS to remove inline scripts and centralize association handling.

ju1ius opened the pull request:
Hi ! :facepunch: 

| Q  |  A |
------------ | -------------
| Branch | master |
| Bugfix? | yes |
| New feature? | yes |
| BC breaks | yes |
| Tests pass? |	yes |
| Fixed tickets | At least: #3646, #3173, #1076, [SonataDoctrineORMAdminBundle/#522](https://github.com/sonata-project/SonataDoctrineORMAdminBundle/issues/522) |

This PR is related to, and inseparable from [SonataDoctrineORMAdminBundle#526](https://github.com/sonata-project/SonataDoctrineORMAdminBundle/pull/526).

## Addressed issues:

* [SonataDoctrineORMAdminBundle/#522](https://github.com/sonata-project/SonataDoctrineORMAdminBundle/issues/522): Adds the missing error handling and progress visual feedback to the AJAX requests.

* #1076: Enhances UX of form validation failures:
When errors are detected in a form: 
  * All tabs containing errors are clearly marked as such.
  * The first tab containing an error is activated.
  * The page is scrolled to the position of the first invalid field label.
  * The offending field is focused.

* #3173 Removes **all** inline, twig-entangled obtrusive javascript from the templates in AdminBundle & SonataDoctrineORMAdminBundle. All data is now passed to javascript as JSON in data attributes, and  event delegation is used whenever possible.

* #3646 Adds PHPUnit assertions properly test Twig's DOM output without relying on it's string representation.

## New Features

* Modularize JS & CSS for maintainability (thanks to webpack, babel, postcss).

* Association CRUD buttons (show/list/edit/create/unlink) are now potentially available to any form type, provided they listen to the right events and handle them correctly. An implementation is currently provided for sonata_type_model, sonata_type_model_list and sonata_type_model_autocomplete. I will also provide an implementation for sonata_type_admin.

* Display list view's show & delete actions in a modal.

* Adds Shift+Click batch selection support for the list views

* Introduces [new, consistently named javascript events](https://github.com/sonata-project/SonataAdminBundle/pull/3647/files#diff-5132598d0fed630b04f7cdace40a0ab0R1) for custom code to hook into.

* Introduces public JS DOM helpers to interact with the sonata ui elements.

* Replaces the unmaintained iCheck javascript library by a CSS only implementation.

* Adds a javascript linting task (eslint)

* Adds an .editorconfig file

* Adds two new classes to the body tag, for easier styling by sonata users. The first class is the snake-cased version of the result of calling `Admin::getCode()` (i.e. `sonata_news_admin_post`). The other is the current action name, prefixed with `action_` (i.e. `action_edit`).

## What's left

### Mongo/PHPCR/Propel
I still haven't updated the other doctrine admin bundles.
The main reason is the following:
Currently the admin bundle delegates to the presistence bundles all the template rendering of the following association form types: `sonata_type_model`, `sonata_type_model_list` and `sonata_type_admin`.
**I strongly question myself about why this is the case at all.**
IMO, it should be part of their contract that these bundles provide view data that is directly consumable by the admin bundle templates.
Maybe we should abstract the mapping type constants (`ClassMetadata::MANY_TO_ONE` et al.) in the ModelManagerInterface ?
Having this separation of concerns would greatly ease future changes.
More importantly, it would allow to test the ui, which can't be done atm since the templates and the logic reside in separate bundles.

The other reason is that the templates are shared between the `sonata_type_model` and `sonata_type_admin` types, and they seem to contain legacy logic to handle the `['edit' => 'list']` option.
This option is no longer allowed on any type other than `sonata_type_admin`, and setting it's value to 'list' turns a `sonata_type_admin` form into something that looks like a `sonata_type_model_list`, except broken.

### Docs
I've made a particular effort to document the code as much as possible, to make future dev & documentation work easier.
When (if) this gets merged, I'll be more than happy to make a PR for updating the docs.

Thanks for taking the time to review this.

ju1ius commented:
I also have the upgrade to select2 v4 in a branch based on this one.
But as this requires a PR to CoreBundle, I'll keep it for later :wink:

ju1ius commented:
Also, if this gets merged, I'd also like to get rid of the annoying confirmExit alert.
Now tha LocalStorage is [widely supported](http://caniuse.com/#search=localstorage), we can easily persist the form state silently and restore it on demand, even after the window/session is closed.

ju1ius commented:
Also, from what I've seen, the whole jQueryUI library is loaded for the sole purpose of using the sortable plugin. If this is the case, then we could remove it in favour of [dragula](https://github.com/bevacqua/dragula), which is much smaller (11Kb vs 228Kb...) and works just as great.

ju1ius commented on Twig/Extension/SonataAdminExtension.php:
```diff
@@ -326,22 +345,162 @@ public function getUrlsafeIdentifier($model, AdminInterface $admin = null)
+        return json_encode($data, JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES);
+    }
+
+    public function renderAutocompleteOptions($context)
```
A lot of this method will be removed once we update to select2 v4, this it allows passing options via html data attributes. So a user can configure the plugin directly using the 'attr' form option.

ju1ius commented on Resources/public/js/Sonata.js:
```diff
@@ -0,0 +1,208 @@
+        useSelect2: true,
+        useICheck: true,
+        useStickyForms: true
+    }, Sonata.config || {});
```
Here we could have one config object per plugin, e.g.
```js
Sonata.config = {
  select2: {enabled: true, defaults: {foo: 'bar'}}
};
```
So users could easily override global config by doing:
```js
Sonata.config.select2.defaults.foo = 'bar';
```

ju1ius commented on Resources/public/js/deprecated/Admin.js:
```diff
@@ -0,0 +1,293 @@
+
+    // For backward compatibility
+    var EVENT_TO_DEPRECATED = {
+        'sonata:append-form-element': 'sonata-admin-append-form-element',
```
forgot to remove that, sorry... :cold_sweat:

ju1ius commented on Resources/public/js/deprecated/Admin.js:
```diff
@@ -0,0 +1,293 @@
+        'pre-delete.sonata.collection': 'sonata-collection-item-deleted',
+        'delete.sonata.collection': 'sonata-collection-item-deleted-successful',
+        'sonata.add_element': 'foo'
+    };
```
forgot to remove that, sorry... :cold_sweat:

Koc commented on Tests/Twig/Extension/SonataAdminExtensionTest.php:
```diff
@@ -283,16 +286,57 @@ public function getRenderListElementTests()
-            array('<td class="sonata-ba-list-field sonata-ba-list-field-boolean" objectId="12345"> <span class="x-editable" data-type="select" data-value="1" data-title="Data" data-pk="12345" data-url="/core/set-object-field-value?context=list&amp;field=fd_name&amp;objectId=12345&amp;code=xyz" data-source="[{value: 0, text: \'no\'},{value: 1, text: \'yes\'}]" > <span class="label label-success">y ...
-            array('<td class="sonata-ba-list-field sonata-ba-list-field-boolean" objectId="12345"> <span class="x-editable" data-type="select" data-value="" data-title="Data" data-pk="12345" data-url="/core/set-object-field-value?context=list&amp;field=fd_name&amp;objectId=12345&amp;code=xyz" data-source="[{value: 0, text: \'no\'},{value: 1, text: \'yes\'}]" > <span class="label label-danger">no< ...
-            array('<td class="sonata-ba-list-field sonata-ba-list-field-boolean" objectId="12345"> <span class="x-editable" data-type="select" data-value="" data-title="Data" data-pk="12345" data-url="/core/set-object-field-value?context=list&amp;field=fd_name&amp;objectId=12345&amp;code=xyz" data-source="[{value: 0, text: \'no\'},{value: 1, text: \'yes\'}]" > <span class="label label-danger">no< ...
+            array(
```
Maybe better use heredoc here?

ju1ius replied:
Yep :ok_hand:

ju1ius commented on Resources/public/js/Sonata.js:
```diff
@@ -0,0 +1,208 @@
+         */
+        getAssociationFieldContainer: function (fieldId, context) {
+            return $('#field_container_' + fieldId, context || document);
+        },
```
Example of bad naming: there's `#sonata-ba-field-container-*` for standard fields, then `#field_container_*` for association fields... :confused:

greg0ire commented on Resources/public/css/components/forms.css:
```diff
@@ -0,0 +1,30 @@
+}
+.nav-tabs .has-error .fa {
+    display: inline-block;
+}
```
Please configure your editor to add a newline at end of file (see that little red pictogram? That's what it means).

greg0ire commented on Resources/public/js/Admin.js:
```diff
@@ -0,0 +1,126 @@
+    /**
+     * Fires when a new empty association form is about to be appended.
+     *
+     * @event 'sonata:association-add'
```
Why quotes there and not line 31. Please pick one.

greg0ire commented on Resources/public/js/deprecated/Admin.js:
```diff
@@ -0,0 +1,280 @@
+        }).done(function (json) {
+            var $parent = $subject.parent();
+            if (json.status === "OK") {
+                // fix issue with html comment ...
```
What issue? This piece of code seems to remove html comments from the json data.

ju1ius replied:
Don't know, the comment was kept from the original code...

greg0ire replied:
That's why I hate "fix bug" comments or commit messages… it makes the code sacred without letting people question if there is indeed a bug and if this is the best way to solve it. Which leads to this : ![genius or stupid](http://www.commitstrip.com/wp-content/uploads/2016/01/Strip-Un-code-inutile-650-finalenglish-1.jpg)

greg0ire commented on Resources/public/js/form/edit-associations.js:
```diff
@@ -0,0 +1,560 @@
+                '</div>' +
+            '</div>' +
+        '</div>'
+    ;
```
Can't the template be loaded from a DOM node? Having html in `.js` does not feel very good…

ju1ius replied:
Honestly, there's absolutely no difference between this and
```js
$('<div class="modal">')
  .append('<div class="modal-content">')
    .append('<div class="modal-header">')
      .append('<button class="close">').attr('aria-hidden', true)
// etc...
```
Except that the first is shorter and more readable...

> Having html in .js does not feel very good…

But it's gonna happen all the time because **that's** what JS is for ! :wink:

greg0ire replied:
You're right, there is no difference, but I was not thinking about that. I was thinking about having a nice syntactically colored html file that could be injected somewhere in the DOM, and then read from it like this (for instance) : 

`template = jQuery('#my-template').html();`

> But it's gonna happen all the time because that's what JS is for ! :wink: 

Is it? I thought JS was here to interact with the DOM, and that html templates should be kept separate… isn't that what ember.js does? IIRC, they have a special attribute for the script tag where they put their templates. The goal would be to have a syntactically colored template one could easily edit without fear of breaking it (meaning having your editor validate it for you).

ju1ius replied:
PHPStorm synax-highlights it very well !

> I thought JS was here to interact with the DOM, and that html templates should be kept separate…

There's been a lot of debate about this recently, with the advent of ReactJS. At the beginning I felt exactly like you (meh...). But after using it extensively, I'm now convinced that having the HTML structure near to the code that use it reduces mental friction and context switches **a lot**.

>  they have a special attribute for the script tag where they put their templates. 

Yeah basically any script tag with a type attribute other that `text/javascript` (or application/javascript at al.) will ignored by the parser.
```html
<script type="application/x-whatever" id="foo">Ignore this</script>
```
But a lot of editors won't syntax-highlight it either. (Vim doesn't with my config)

Koc replied:
You can use `script` tag for storing templates. Browsers are not parse html inside script tag. `<script type="text/html">...`

ju1ius replied:
To sum up I'm not against using the solution you mentioned (custom script tags) or another (like a js/templates directory with html files loaded via ajax, or whatever).

I just think that currently there's no actual need for this because the codebase is so small.
In fact if they were numerous templates I probably would have done it.

greg0ire replied:
I think this can lead to easy coloration / validation in all editors.

`<script type="application/x-whatever" id="foo">{{ include 'nicely-colored-template.html' }}</script>`

> But after using it extensively, I'm now convinced that having the HTML structure near to the code that use it reduces mental friction and context switches a lot.

Just like annotation do reduce mental friction. But we must be careful with this, when the code grows or needs more flexibility / decoupling it often ends up being a wrong choice. Anyway, don't bother if there is not that many templates, but if the number of template grows, it could be worth considering the switch to a better solution.

greg0ire commented on Resources/public/js/form/edit-associations.js:
```diff
@@ -0,0 +1,560 @@
+    });
+
+    // Used in: edit_orm_one_to_many, edit_orm_many_to_many
+    $(document).on('click.sonata-admin', '.sonata-ba-action[data-field-action="append-form-element"]', function (event) {
```
Please linebreak lines 538 and 554

greg0ire commented on Resources/public/js/form/errors.js:
```diff
@@ -0,0 +1,223 @@
+            if (tabHasError(tab)) {
+                tab.trigger.parent().addClass('has-error');
+            } else {
+                tab.trigger.parent().removeClass('has-error');
```
Could be written in a more concies manner with `toggleClass()` : 

`tab.trigger.parent().toggleClass('has-error', tabHasError(tab));`

greg0ire commented on Resources/public/js/form/errors.js:
```diff
@@ -0,0 +1,223 @@
+                input.setAttribute('aria-invalid', true);
+            } else {
+                input.setAttribute('aria-invalid', false);
+            }
```
`input.setAttribute('aria-invalid', $(input).closest('.sonata-ba-field-error').length > 0)`

greg0ire commented on Resources/views/CRUD/base_edit_form.html.twig:
```diff
@@ -35,13 +37,21 @@
                                 {% for code, form_tab in admin.formtabs %}
-                                    <div class="tab-pane fade{% if loop.first %} in active{% endif %}" id="tab_{{ admin.uniqid }}_{{ loop.index }}">
-                                        <div class="box-body  container-fluid">
+                                    <div id="tab_{{ admin.uniqid }}_{{ loop.index }}" class="{{ classlist('tab-pane', 'fade', {
```
Please linebreak this

greg0ire commented on Resources/views/CRUD/base_history.html.twig:
```diff
@@ -33,16 +33,35 @@ file that was distributed with this source code.
-                            <tr class="{% if (currentRevision != false and revision.rev == currentRevision.rev) %}current-revision{% endif %}">
+                            <tr class="{{ classlist({
+                                'sonata-ba-revisions__revision': true,
+                                'sonata-ba-revisions__revision--is-current': currentRevision != false and revision.rev == currentRevision.rev
```
Linebreak here too please

ju1ius replied:
Linebreak where ? how ?

greg0ire replied:
Where : as soon as you need to use the horizontal scrollbar on github would be great. Having lines shorter than 120 chars should be the goal IMO.

How : in this instance, like this maybe ? :  
                                'sonata-ba-revisions__revision--is-current': 
                                    currentRevision != false 
                                    and 
                                    revision.rev == currentRevision.rev

ju1ius replied:
Okay so let's say soft limit at 80 and hard limit at 120 ?

greg0ire replied:
Exactly :) 80 is very hard with that level of indention, but if you're in an unindented file, try to achieve it.

greg0ire commented on Resources/views/CRUD/base_history.html.twig:
```diff
@@ -33,16 +33,35 @@ file that was distributed with this source code.
-                                <td>{{ revision.username|default('label_unknown_user'|trans({}, 'SonataAdminBundle')) }}</td>
-                                <td><a href="{{ admin.generateObjectUrl('history_view_revision', object, {'revision': revision.rev }) }}" class="revision-link" rel="{{ revision.rev }}">{{ "label_view_revision"|trans({}, 'SonataAdminBundle') }}</a></td>
+                                <td>
+                                    {{- revision.username|default('label_unknown_user'|trans({}, 'SonataAdminBundle')) -}}
```
linebreak here too, or please consider using `trans_default_domain`

greg0ire commented on Resources/views/Form/form_admin_fields.html.twig:
```diff
@@ -294,26 +301,27 @@ file that was distributed with this source code.
+                        {% set grid_class = 'col-sm-12' %}
+                    {% else %}
+                        {% set grid_class = 'col-sm-9 col-sm-offset-3' %}
+                    {% endif %}
```
Consider using a ternary.

greg0ire commented on Resources/views/standard_layout.html.twig:
```diff
@@ -32,30 +33,34 @@ file that was distributed with this source code.
+                    confirmExit: {{ sonata_admin.adminPool.getOption('confirm_exit') ? 'true' : 'false' }},
+                    useSelect2: {{ sonata_admin.adminPool.getOption('use_select2') ? 'true' : 'false' }},
+                    useICheck: {{ sonata_admin.adminPool.getOption('use_icheck') ? 'true' : 'false' }},
+                    useStickyForms: {{ sonata_admin.adminPool.getOption('use_stickyforms') ? 'true' : 'false' }}
```
How about using `Boolean.toString();` ?

ju1ius replied:
I personally prefer
```twig
{{ sonata_admin.adminPool.getOption('confirm_exit')|json_encode }}
```
But I wasn't 100% shure what the actual value could be...

greg0ire replied:
Oh yeah sorry, this is twig, not javascript, my bad.

greg0ire commented on Tests/Twig/Extension/SonataAdminExtensionTest.php:
```diff
@@ -247,7 +248,7 @@ public function testRenderListElement($expected, $type, $value, array $options)
             }));
 
-        $this->assertSame($expected, trim(preg_replace('/\s+/', ' ', $this->twigExtension->renderListElement($this->object, $this->fieldDescription))));
+        $this->assertHTMLequals($expected, $this->twigExtension->renderListElement($this->object, $this->fieldDescription));
```
please linebreak this.

greg0ire commented on Tests/Twig/Extension/SonataAdminExtensionTest.php:
```diff
@@ -263,6 +264,7 @@ public function getRenderListElementTests()
             array('<td class="sonata-ba-list-field sonata-ba-list-field-textarea" objectId="12345"> </td>', 'textarea', null, array()),
             array('<td class="sonata-ba-list-field sonata-ba-list-field-datetime" objectId="12345"> December 24, 2013 10:11 </td>', 'datetime', new \DateTime('2013-12-24 10:11:12', new \DateTimeZone('Europe/London')), array()),
             array('<td class="sonata-ba-list-field sonata-ba-list-field-datetime" objectId="12345"> December 24, 2013 18:11 </td>', 'datetime', new \DateTime('2013-12-24 10:11:12', new \DateTimeZone('UTC')), array('timezone' => 'Asia/Hong_Kong')),
+            // #10
```
A `#`, and not for 20 or 30 ?

greg0ire commented on Tests/Twig/Extension/SonataHelpersExtensionTest.php:
```diff
@@ -0,0 +1,90 @@
+            // classlist('foo') is a shortcut to classlist({foo: true})
+            array(array('foo', 'bar'), 'foo bar'),
+            array(array(array('foo' => true, 'bar' => true)), 'foo bar'),
+            // ignores falsy values
```
Instead of using comments, you can use them directly as the indices of the array : If the test fails for a given index, the string will appear in the error message :)

ju1ius replied:
WOW I didn't know that, great !

greg0ire replied:
Learnt it here if you're interested : http://fr.slideshare.net/javier.eguiluz/new-symfony-tips-tricks-symfonycon-paris-2015 This talk is a goldmine

greg0ire commented on Tests/Twig/HTMLTestCase.php:
```diff
@@ -0,0 +1,153 @@
+            $message .= "\n";
+        }
+        $this->assertSame($expected->tagName, $actual->tagName, sprintf(
+            '%sTag name doesn\'t match',
```
Please use double quotes to avoid escaping.

greg0ire commented on Twig/Extension/SonataAdminExtension.php:
```diff
@@ -76,6 +78,23 @@ public function getFilters()
+                'needs_context' => true,
+                'is_safe'       => array('html'),
+            )),
+            new \Twig_SimpleFunction('sonata_autocomplete_options', array($this, 'renderAutocompleteOptions'), array(
```
Linebreaks for 84 and 88 please.

greg0ire commented on Twig/Extension/SonataAdminExtension.php:
```diff
@@ -326,22 +345,162 @@ public function getUrlsafeIdentifier($model, AdminInterface $admin = null)
+            'associationAdmin' => $associationAdmin->getCode(),
+            'formType'         => $fieldDescription->getType(),
+            'editMode'         => $sonataAdmin['edit'],
+            'label'            => $associationAdmin->trans($associationAdmin->getLabel(), array(), $associationAdmin->getTranslationDomain()),
```
linebreak here too please

ju1ius replied:
Linebreaking + aligned arrays === seasick ...

greg0ire replied:
Yeah, about aligned arrays, not sure if alignment is enforced, but if it isn't, just don't do it. Aligning things together is coupling lines that do not related, and makes diffs harder to read.

ju1ius replied:
It's enforced in `.php_cs`. If you don't do it, travis not happy...

ju1ius replied:
But I'd be happy to change that :grin:

greg0ire replied:
@Soullivaneuh, what do you think ?

Soullivaneuh replied:
Stop me if I'm wrong, but you can do something like that:

```php
$associationAdmin->trans(
    $associationAdmin->getLabel(),
    array(),
    $associationAdmin->getTranslationDomain()
),
```

Without having php cs yelling... Check also if you are using latest version of php-cs-fixer.

ju1ius replied:
I was talking about the `align_double_arrow` option.
Once you start linebreaking a method call in an aligned array, things start looking weird....
```php
$config = array(
    'id'               => $context['id'],
    'label'            => $associationAdmin->trans(
        $associationAdmin->getLabel(),
        array(),
        $associationAdmin->getTranslationDomain()
    ),
    'select2' => array(
        'dropdownCssClass'   => $context['dropdown_css_class'],
        'ajax'               => array(
            'url'         => $context['url'] ?: $router->generate(
                $context['route']['name'],
                $context['route']['parameters']
            ),
            'quietMillis' => $context['quiet_millis'],
            'cache'       => $context['cache'],
        ),
    ),
);
```

ju1ius replied:
.php_cs
```php
return Symfony\CS\Config\Config::create()
    ->level(Symfony\CS\FixerInterface::SYMFONY_LEVEL)
    ->fixers(array(
        'header_comment',
        '-unalign_double_arrow',
        '-unalign_equals',
        'align_double_arrow',
        'newline_after_open_tag',
        'ordered_use',
        'long_array_syntax',
        'php_unit_construct',
        'php_unit_strict',
    ))
    ->setUsingCache(true)
    ->finder($finder)
;
```

greg0ire replied:
@rande : can we change this `-unalign_double_arrow` rule ? @Soullivaneuh tells me he doesn't feel very good about it either, but that I should see that with you. If yes, should we change it to `unalign_double_arrow` ? Or remove it ?

greg0ire replied:
One more argument against alignment : conflicts. The number of conflicts will grow with the size of the array. Sorting the array and avoiding alignment is the best way to avoid conflicts in big arrays.

greg0ire commented on Twig/Extension/SonataAdminExtension.php:
```diff
@@ -326,22 +345,162 @@ public function getUrlsafeIdentifier($model, AdminInterface $admin = null)
+
+        $allowClearPlaceholder = (!$multiple && !$required) ? ' ' : '';
+
+        $data = array(
```
Can you please find a better variable name than `$data`? `$autocompleteOptions`, maybe ?

ju1ius replied:
mmmh, in fact that's not really options, since they are required...
maybe parameters ?

greg0ire replied:
The method is named `renderAutocompleteOptions`, that's why I was suggesting that : it should return auto-complete options. Gotta find a name that fits this big array of data that are somewhat related anyway. `$autocompleteAttributes, maybe` ? Or `$autcompleteConfiguration` ? It should convey the following (correct me if I'm wrong) : 

- this is about configuring a widget
- the widget is an autocomplete widget

ju1ius replied:
> $autcompleteConfiguration
:+1:  yessss, I'll change the method name too

ju1ius replied:
Aaaargh it makes me want to name it `$config` now !
I must have a coffee ! :coffee: :coffee: :coffee:

greg0ire replied:
`$config` would be ok, do what you feel is right.

greg0ire commented on Twig/Extension/SonataAdminExtension.php:
```diff
@@ -326,22 +345,162 @@ public function getUrlsafeIdentifier($model, AdminInterface $admin = null)
+        $multiple = $context['multiple'] ? true : false;
+        $required = $context['required'] ? true : false;
+
+        $allowClearPlaceholder = (!$multiple && !$required) ? ' ' : '';
```
Equivalent to the less negative `($multiple || $required) ? '' : ' '`

greg0ire commented on Twig/Extension/SonataHelpersExtension.php:
```diff
@@ -0,0 +1,97 @@
+    public function getFunctions()
+    {
+        return array(
+            new \Twig_SimpleFunction('classlist', array($this, 'renderClassList'), array('is_safe' => array('html_attr'))),
```
Linebreak please.

greg0ire commented:
Stellar work @ju1ius ! This looks really nice!!! If you are following some standards, (and it seems you are), would you please document them in CONTRIBUTING.md ? So that other contributors stay consistent with what you did?

Koc commented:
> Also, if this gets merged, I'd also like to get rid of the annoying confirmExit alert.
> Now that LocalStorage is widely supported, we can easily persist a form's state silently when the user goes away, and ask him later if he wants his unsaved changes back.

Please not. Forms can be too big, also you should store data for edit form for each object that we opening. And also it can be inconsistency with form data on the server and form data in LS. Also it is some edge cases like ignoring storing csrf `_token`

Koc commented:
Dear Sonata members, is this PR can be merged on 2.4 version or only on 3 because of BC breaks on twig and js?

core23 commented:
I think we have to wait for 3.x to get this :( There are many breaks which overlap with the other Sonata*AdminBundles

ju1ius commented:
> Please not. Forms can be too big, also you should store data for edit form for each object that we opening. 
@Koc Currently the form state is saved by serializing it to json and storing it in a `data-original` attribute on the form element. I fail to see in what way is this better than saving to localStorage.

greg0ire commented:
> I think we have to wait for 3.x to get this 

Ok, but please do not let this PR rot / become incompatible with other things. If there is a feature freeze, then BC-breaking changes should be merged anyway, but in a new, dedicated branch. If we wait for 2.4 to be out to start merging 3.0 new features, we will have a conflict nightmare, and contributors will need to rebase, and some of then will be unreachable / deceased… What do you think?

ju1ius commented:
@core23 Would there be a way for it to be backward compatible ? I don't know the whole codebase enough...

Koc commented:
@core23 can we force 3.x or maybe do not release 2.4? @ju1ius have done a big work and it is not possible rebease too much code all the time when something will cnahge

greg0ire commented:
ping @Soullivaneuh, we are talking release process here, it might interest you.

ju1ius commented:
Concerning BC:
Many dynamic languages do not support the public/protected/private distinction.
In Python, the convention is that anything that is not _documented_ is not part of the public API. Things like prefixing variables and methods with an underscore are just considered as hints to the developer to use it at their own risk. So as the javascript API was not documented, one might say there's technically no BC break.... And since we're just initializing some plugins, [this might be enough](https://github.com/sonata-project/SonataAdminBundle/pull/3647/files#diff-bf92b9c1651d01b45467998006f1f271R193) for keeping BC.

Changing the twig templates is more problematic though, since people could rely on the html structure for their custom code. (I do).
And upgrading to select2 v4 will definitely break BC since there will be changes in the `ModelAutocompleteType` class.

ju1ius commented:
>  If you are following some standards, (and it seems you are), would you please document them in CONTRIBUTING.md ? So that other contributors stay consistent with what you did?

There are tools for that ([eslint](http://eslint.org/), [jscs](http://jscs.info/)) I'd be happy to add them, but we need to switch from bower to npm...

ju1ius commented:
> also you should store data for edit form for each object that we opening.

@Koc  No, you would listen to `window.onvisibilitychange` and save the current form(s) when state becomes either 'hidden' or 'unloaded'. See https://www.igvita.com/2015/11/20/dont-lose-user-and-app-state-use-page-visibility/

greg0ire commented:
> There are tools for that (eslint, jscs) I'd be happy to add them, but we need to switch from bower to npm...

There is a `cs` target in the Makefile, having a `js-cs` target based on what you recommend would be great!

Soullivaneuh commented:
>  but we need to switch from bower to npm...

Not necessary switch. We can use both npm and bower.

ju1ius commented:
> Not necessary switch. We can use both npm and bower.

Indeed!

> There is a cs target in the Makefile, having a js-cs target based on what you recommend would be great!

I'll have a look at that.

ju1ius commented:
For the JS part, one way to be 100% BC:
Say we use babel+webpack,
```js
// In public/js/plugins/my-plugin.js
const {jQuery} = window;

export default function launchZeMissiles ($element) {
  // actually launch teh missilez
}
jQuery($ => launchZeMissiles($(document)));
jQuery(document).on('sonata:domready', e => launchZeMissiles(jQuery(e.target)));
```

```js
// In public/js/Admin.js
import launchZeMissiles from 'plugins/my-plugin';

const Admin = {
  setup_my_plugin (subject) {
    console.warn('Deprecated in version 2.4, to be removed in 3.0. Use something else instead');
    launchZeMissiles(subject);
  }
};
window.Admin = Admin;
```

This way we don't expose `launchZeMissiles` publicly, and allow the old API to be used during transition phase.

ju1ius commented:
> I think we have to wait for 3.x to get this :( There are many breaks which overlap with the other Sonata*AdminBundles

@core23 Are you sure ?
Please do correct me if I'm wrong:
* From the PHP point of view there's no BC break.
* The javascript **can** be made BC with a little bit of work, [see the above comment](#issuecomment-196012041) (there are other solutions not involving es2015 modules, although not as pretty).
* Changes to the templates are not so big. There's no change in the HTML structure, just in some classes/attributes, and of course the inline scripts removed in the [doctrine bundles](https://github.com/sonata-project/SonataDoctrineORMAdminBundle/pull/526).
* Browsing the repo history since 2.3, I see many changes made to the templates that did not deserve a new _major_ version.

Concerning the other bundles:
```sh
$ cd sonata/sandbox
$ git rev-parse --abbrev-ref HEAD
> 2.4-develop
$ cd vendor/sonata-project
# search for twig or unminified javascript files,
$ ag -G '.*((?<!\.min)\.js|\.twig)$' \
    # ignoring the admin-bundle directory and showing only the matching parts,
    --ignore 'admin-bundle' --only-matching \
    # containing calls to the old js api
    '\bAdmin\.\w+\('
> media-bundle/Resources/views/MediaAdmin/edit.html.twig
> 137:Admin.setup_list_modal(
>
> page-bundle/Resources/assets_src/src/js/composer.js
> 59:Admin.shared_setup(
>
> page-bundle/Resources/public/sonata-page.back.js
> 72:Admin.shared_setup(
```

So it looks like once BC is provided for the old `Admin` methods, everything will work perfectly fine in all the other feature bundles.
And there will be only 2 small PRs to submit to update media-bundle and page-bundle.

ju1ius commented:
I also forgot to mention that among these changes, there's also Shift+Click batch selection support for the list views...

core23 commented:
> Many dynamic languages do not support the public/protected/private distinction.
In Python, the convention is that anything that is not documented is not part of the public API. Things like prefixing variables and methods with an underscore are just considered hints to the developer to "use this at your own risks". So as the javascript API was not documented, one might say there's technically no BC break.... And since we're just initializing some plugins, this might be enough for keeping BC.

I know. Symfony uses something called **api** methods which should stay BC and couldn't be modified. I tried to talk to some other devs about this, but it's kind a hard because there exists no offical API stuff because everything is API. Maybe we should start working on the next major release after the next minor (2.4) to get rid of many old code.

> Changes to the templates are not so big. There's no change in the HTML structure, just in some classes/attributes, and of course the inline scripts removed in the doctrine bundles.

You removed some templates, which is a BC break (see https://github.com/sonata-project/SonataDoctrineORMAdminBundle/pull/526)

> Browsing the repo history since 2.3, I see many changes made to the templates that did not deserve a new major version.

That's right. In the last years, there have been some violations against this rule, but we all would become more responsible and try to respect http://semver.org/.

greg0ire commented:
> I know. Symfony uses something called api methods which should stay BC and couldn't be modified. I tried to talk to some other devs about this, but it's kind a hard because there exists no offical API stuff because everything in the sonata bundles is API today. Maybe we should start working on the next major release after the next minor (2.4) to get rid of many old code.

Maybe we should also use more final classes / private methods

related : 

- https://ocramius.github.io/blog/when-to-declare-classes-final/
- http://fabien.potencier.org/pragmatism-over-theory-protected-vs-private.html

ju1ius commented:
> Maybe we should start working on the next major release after the next minor (2.4) to get rid of many old code.

For the JS code, the most important would be to document what's public, so that you can officially state that the rest can change.

> You removed some templates, which is a BC break

This template thing is annoying. In the symfony system, as soon as you ship a template in a bundle it can be overriden or extended, am I right ? So technically any change to any template is a potential BC break since anybody can start relying on a particular DOM structre or attribute or whatever. :scream: 

> we all would become more responsible and try to respect http://semver.org/.

In which case 2.4 should be 3.0 ?

ju1ius commented:
As a Sonata user, I rely a lot on custom code an overriden templates. Given stable updates are scarce, I'm almost always depending on dev-master, then each time I make a `composer update`, I make it in a seperate branch, make a `git --dirdiff` of each sonata bundle and look for things that might break my app.

From my POV, I would largely prefer that the templates can change between minor releases (as long as the blocks structure stays the same), and having more official releases with a changelog. Even just a list of filenames that have been touched would be a time-saver.

ju1ius commented:
> You removed some templates, which is a BC break

@core23 So if I put back these templates in place with nothing but whitespace in them, it's non-BC ?

ju1ius commented:
@core23 @Soullivaneuh and all others, I'm really concerned about BC, because 2.3 was released a year and a half ago now, the 2.4 milestone is at 70% and the 3.0 branch doesn't exist, so _anything I'll have to do I'll do_.

greg0ire commented:
@ju1ius : it is only a matter of time before @rande and @Soullivaneuh finalize setting up #3053 , but as you can see, things are supposed to change soon, in the good direction.

greg0ire commented:
@ju1ius please check your laposte.net inbox, I invited you on slack so that we can discuss things like this more easily :)

ju1ius commented:
Ouch, travis's node version is 0.10...

ju1ius commented:
Hi,
so, I updated this PR and [SonataDoctrineORMAdminBundle#526](https://github.com/sonata-project/SonataDoctrineORMAdminBundle/pull/526) to add BC back (among other things). :coffee: 
It's almost done now, I just need to add-back the old three jQuery events on forms.
I'll update the main PR description in order to limit the comment count.

ju1ius commented:
@core23 @Soullivaneuh Okay, as stated above, I added the legacy events back, so considering this and the changes in [SonataDoctrineORMAdminBundle#526](https://github.com/sonata-project/SonataDoctrineORMAdminBundle/pull/526) it should be near 100% BC now.

Please have a look at this and tell me if more things have to be done.

Cheers.

greg0ire commented on .travis.yml:
```diff
@@ -42,6 +43,10 @@ matrix:
     - php: hhvm
 
+install:
+  - rm -rf ~/.nvm && git clone https://github.com/creationix/nvm.git ~/.nvm && (cd ~/.nvm && git checkout `git describe --abbrev=0 --tags`) && source ~/.nvm/nvm.sh && nvm install $TRAVIS_NODE_VERSION
```
Please : 
- linebreak this
- use long options.
- use `$()` instead of backticks (they are deprecated)

ju1ius replied:
> linebreak this

I was afraid of a make-like behavior where you loose the cwd on each line. Is this the case ?

> use long options.

Which ones ?

> use $() instead of backticks (they are deprecated)

Okay so this is a bash shell. Good to know.

greg0ire replied:
Sorry, I forgot to answer.

> I was afraid of a make-like behavior where you loose the cwd on each line. Is this the case ?

Every language has a way to avoid long lines (I think). For YAML, you can do it like this : https://github.com/greg0ire/enum-bundle/blob/master/.travis.yml#L19

> Which ones ?

Every one you can. Please consider that short options are for you, in your shell, but as soon as you write an option in a script or in documentation, you must be as noob-compliant (meaning explicit) as you can. Maybe this is a bit extreme for `rm -rf`, that can be considered well-known, but since you now know how to avoid long line, it is not really a problem to use `--recursive --force` instead.

ju1ius replied:
@greg0ire Wouldn't it be better to have a something like a `.travis` directory and put shell scripts inside ?

greg0ire replied:
A separate directory is a good idea, but maybe don't call it travis but something more generic, like `ci`.

greg0ire commented on Form/Type/ModelTypeList.php:
```diff
@@ -47,6 +47,7 @@ public function buildView(FormView $view, FormInterface $form, array $options)
             $view->vars['sonata_admin']['edit'] = 'list';
         }
         $view->vars['btn_add'] = $options['btn_add'];
+        $view->vars['btn_show'] = $options['btn_show'];
```
Maybe it is time to refactor this, it is getting a bit repetitive.

greg0ire commented on Resources/assets/css/layout/sidebar.css:
```diff
@@ -0,0 +1,39 @@
+
+.nav-third-level li a {
+    padding-left: 52px;
+}
```
Looks like the corresponding code was removed in f5383def0323f1796125bc6580ec1836127083c1

greg0ire commented on Resources/assets/css/legacy.css:
```diff
@@ -0,0 +1,68 @@
+}
+
+/*
+ * The .filter_legend class isn't used anywhere
```
It appears in `Resources/views/CRUD/base_list.html.twig`

ju1ius replied:
mmmh, I can't find it in my working copy....
Do you have a link ?

greg0ire replied:
My bad, I found it on 2.3, it was removed in 0f67eb74491e083b93bc7f3e3de4a3de4b9ca661

ju1ius replied:
I must learn to use the git log...

greg0ire replied:
:) Here is what I did : `git log -S filter_legend -p Resources/views/CRUD/base_list.html.twig`

OskarStark commented:
what is the current status of this PR?

ju1ius commented:
@OskarStark I'm currently using it in a private project. I've fixed a few bugs, added a few features and rebased regularly on master, but haven't pushed them yet. 
I'm waiting because I had no response concerning the possibility of having it merged into 2.4...
Trying to maintain BC means more work, so I'll be happy to drop the compat layer if this must go in 3.0.

@Soullivaneuh When will the 3.0 branch be opened ?

greg0ire commented:
> @Soullivaneuh When will the 3.0 branch be opened ?

Never, we will open 2.x instead, supposedly this week-end (right @Soullivaneuh ?) and master will be aliased to 3.0 .

ju1ius commented:
> Never, we will open 2.x instead, supposedly this week-end (right @Soullivaneuh ?) and master will be aliased to 3.0.

Great! I'll update this PR ASAP.
I'm gonna drop the BC for 2.4 since nobody seems interested.

greg0ire commented:
>  I'm gonna drop the BC for 2.4 since nobody seems interested.

I think for great refactorings like that, it's best to drop BC, because it's hard to actually be sure you keep it. Can you document the BC-breaks in a file named `UPGRADE/2-to-3.md` ?

ju1ius commented:
@greg0ire @OskarStark @core23 @Soullivaneuh @Koc Updated, rebased, squashed, etc...

greg0ire commented:
I'll try running the build again, the error you are getting is strange.

greg0ire commented:
Looks like I don't have the right to do that, nothing happens when I click "retry".

OskarStark commented:
wow this is a huge change, thanks for this effort! :clap: :beers: 

i need some time to review this, maybe @mvhirsch can support us!

greg0ire commented:
@OskarStark can you retry the build ? Looks like I can't.

ju1ius commented:
@OskarStark @greg0ire  I just saw I few changes I have to make so I'll push again in a few mins

greg0ire commented:
Ok great.

ju1ius commented:
@OskarStark Should I update https://github.com/sonata-project/SonataDoctrineORMAdminBundle/pull/526 and remove the templates ?

core23 commented:
This pr changes many many files... Can you try to separate some stuff to a seperate commit? Maybe global config files?

ju1ius commented:
> Maybe global config files?

@core23  Why do you mean by that ?

ju1ius commented:
@core23 are you talking about .travis.yml, Makefile & .editorconfig ?

ju1ius commented:
> Can you try to separate some stuff to a seperate commit?

Separate commit or PR ? Separate PR is gonna be very hard. Separate commit not so much.

greg0ire commented on .editorconfig:
```diff
@@ -0,0 +1,18 @@
+indent_size = 2
+
+[Makefile]
+indent_style = tab
```
This should indeed be in a separate commit. It deals with many things, not just JS. Should be labelled "provide interoperable editor configuration", for instance

ju1ius replied:
> This should indeed be in a separate commit.

Commit or PR ?

greg0ire replied:
Commit, no need to make it too hard…

Soullivaneuh replied:
@ju1ius Can you please remove it? It will be auto updated by dev-kit later: https://github.com/sonata-project/dev-kit/blob/master/tools/editorconfig/.editorconfig

Feel free to contribute on the following link if necessary. :+1:

ju1ius replied:
Yes I'll keep an eye on master and remove it as soon as you commit it.

greg0ire commented on .php_cs:
```diff
@@ -27,7 +27,7 @@ return Symfony\CS\Config\Config::create()
         'long_array_syntax',
         'php_unit_construct',
-        'php_unit_strict',
+        'php_unit_strict'
```
You did not intend to change that, I guess.

ju1ius replied:
Indeed

greg0ire commented on .eslintrc.js:
```diff
@@ -0,0 +1,98 @@
+        "prefer-template": 2,
+        "template-curly-spacing": [2, "never"]
+    }
+};
```
This could also be in a separate commit, since it does not have much to do with the js refactoring (maybe it depends on it, but it is another concern nonetheless).

greg0ire commented on Makefile:
```diff
@@ -1,14 +1,18 @@
 cs:
 	php-cs-fixer fix --verbose
+	npm run lint -- --fix
```
If i advise someone to run make cs to fix the build, and they don't have npm (and they don't care because they don't work on the JS part), things are going to get complicated. How about :

```
backend_cs:
	php-cs-fixer fix --verbose

backend_cs_dry_run:
	php-cs-fixer fix --verbose --dry-run

frontend_cs:
	npm run lint -- --fix

frontend_cs_dry_run:
	npm run lint

cs: backend_cs frontend_cs

cs_dry_run: backend_cs_dry_run frontend_cs_dry_run
```

?

ju1ius replied:
> and they don't have npm

Who doesn't ? :laughing:

greg0ire replied:
> > and they don't have npm

> Who doesn't ? :laughing:

I think you would be amazed.

ju1ius commented:
Jenkins says merge conflicts? :confused:

core23 commented:
IMO separate commits could speed up the review process

ju1ius commented:
@core23 I split as @greg0ire suggested in [his](https://github.com/sonata-project/SonataAdminBundle/pull/3647/files#r58581820) [comments](https://github.com/sonata-project/SonataAdminBundle/pull/3647/files#r58580708).

core23 commented on Makefile:
```diff
@@ -1,14 +1,26 @@
+
+cs_dry_run: phpcs_dry_run eslint_dry_run
+
+test: phpunit docs
```
:-1: Please revert this. Docs build should only run once in travis

ju1ius replied:
@core23 Yep, I guess git was not able to magically merge your commit !

greg0ire commented:
> >     I'm gonna drop the BC for 2.4 since nobody seems interested.
>
> I think for great refactorings like that, it's best to drop BC, because it's hard to actually be sure you keep it. Can you document the BC-breaks in a file named UPGRADE/2-to-3.md ?

Looks like there is another convention : `UPGRADE-3.0.md`

ju1ius commented:
@greg0ire @OskarStark @Soullivaneuh @core23 @rande 
Hi,
I'm currently refactoring this PR to allow developers to add the association crud buttons to any type of widget without having to re-implement the whole ajax/dialog logic.
While working on it, I found that when you add a `sonata_type_admin` field with the 'edit' option set to 'list':
```php
// ManyToOne
$formMapper->add('location', 'sonata_type_admin', [], ['edit' => 'list']);
```
The resulting UI is almost exactly the same as `sonata_type_model_list`. The difference is that the related admins fields are embedded but hidden inside a `<div style="display:none>` element...

To me this doesn't make sense since the whole purpose of the admin type is to edit related objects inline.

So I'm wondering:
* What was the reasoning behind this ?
* Isn't it just the remainings of a legacy implementation that can be removed, [just as it has been for `sonata_type_model`](https://github.com/sonata-project/SonataDoctrineORMAdminBundle/blob/master/Builder/FormContractor.php#L90) ?

Thanks for answering.

OskarStark commented on Form/Type/ModelTypeList.php:
```diff
@@ -68,12 +69,13 @@ public function setDefaultOptions(OptionsResolverInterface $resolver)
+            'btn_show'       => 'link_action_show',
+            'btn_list'       => 'link_list',
+            'btn_delete'     => 'link_delete',
+            'btn_catalogue'  => 'SonataAdminBundle',
```
why did you add an additional whitespace before the `=>` ?

ju1ius replied:
The make cs did that...

ju1ius replied:
How I'd love to ditch that align_double_arrow thing... :grin:

OskarStark commented on Makefile:
```diff
@@ -1,14 +1,26 @@
 
-cs_dry_run:
+eslint:
+	npm run lint -- --fix
```
`-- --fix` is correct?

greg0ire replied:
`--` is often used to separate args and options, so it might be.

ju1ius replied:
> -- --fix is correct?

Yes. `npm run lint -- --fix` tells npm to run the "lint" script from the package.json. so the resulting command is `./node_modules/.bin/eslint "--fix"`

OskarStark commented on Resources/translations/SonataAdminBundle.en.xliff:
```diff
@@ -202,6 +202,14 @@
             </trans-unit>
+            <trans-unit id="label_list_batch_select">
+                <source>label_list_batch_select</source>
+                <target>Select / deselect all rows</target>
```
german translation:
`Alle Zeilen anwählen / abwählen`

@core23 what would you say?

OskarStark commented on Resources/translations/SonataAdminBundle.en.xliff:
```diff
@@ -202,6 +202,14 @@
+            </trans-unit>
+            <trans-unit id="label_list_select_row">
+                <source>label_list_select_row</source>
+                <target>Select / deselect this row</target>
```
german translation:
`Zeile anwählen / abwählen`

@core23 what would you say?

OskarStark commented on Tests/HTMLAssert.php:
```diff
@@ -0,0 +1,199 @@
+     *   - Whitespace is normalized.
+     *   - Attributes names & values are checked (order doesn't matter).
+     *   - The order of classes in the class attribute doesn't matter.
+     *
```
1 blank line is enough

OskarStark commented on Tests/HTMLAssert.php:
```diff
@@ -0,0 +1,199 @@
+    public static function assertHTMLElementEquals(\DOMElement $expected, \DOMElement $actual, $message = '')
+    {
+        if ($message) {
+            $message .= "\n";
```
what about using PHP_EOL constant here?

OskarStark commented on Tests/HTMLAssert.php:
```diff
@@ -0,0 +1,199 @@
+        $node->normalize();
+    }
+
+    private static function dumpHtmlNode(\DOMNode $element, $deep = false)
```
phpdoc missing

ju1ius replied:
Is it required for private ?

OskarStark commented on Tests/Twig/Extension/SonataAdminExtensionTest.php:
```diff
@@ -126,6 +128,8 @@ public function setUp()
         $this->environment->addExtension(new RoutingExtension($urlGenerator));
         $this->environment->addExtension(new \Twig_Extensions_Extension_Text());
 
```
would remove the blank line here to group all extensions

OskarStark commented on Twig/Extension/SonataAdminExtension.php:
```diff
@@ -391,22 +426,199 @@ public function getUrlsafeIdentifier($model, AdminInterface $admin = null)
+        return json_encode($config, JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES);
+    }
+
+    public function renderAutocompleteConfiguration($context)
```
php doc missing

OskarStark commented on Twig/Extension/SonataHelpersExtension.php:
```diff
@@ -0,0 +1,146 @@
+    {
+        return array(
+            new \Twig_SimpleFunction(
+                'classlist',
```
what about `class_list` or `sonata_class_list` ?

ju1ius replied:
I thought using the DOM API name was cool... and short.

OskarStark commented:
@ju1ius i added some comments!

btw this PR is awesome 👏

ju1ius commented:
@OskarStark Thanks for your input, I updated the PR with your comments, and more !

ju1ius commented:
@greg0ire  @Soullivaneuh @core23  I refactored this PR a bit.
The main important change is, as stated in the edited main PR comment:

Association CRUD buttons (show/list/edit/create/unlink) are now potentially available to any form type, provided they listen to the right events and handle them correctly.
An implementation is currently provided for sonata_type_model, sonata_type_model_list and sonata_type_model_autocomplete.

I will also provide an implementation for sonata_type_admin.
To do this, however I still need an answer (or any kind of input really, even a global IDKWTFYTA will be helpful) [to my previous question](https://github.com/sonata-project/SonataAdminBundle/pull/3647#issuecomment-211441695).

I also have another itch to scratch:
Currently the admin bundle delegates to the persistence bundles all the template rendering of the following association form types: sonata_type_model, sonata_type_model_list and sonata_type_admin.
**I strongly question myself about why this is the case at all.**
IMO, it should be part of their contract that these bundles provide view data that is directly consumable by the admin bundle templates.
Maybe we should abstract the mapping type constants (ClassMetadata::MANY_TO_ONE et al.) in the ModelManagerInterface ?

Having this separation of concerns would greatly ease future changes.
More importantly, it would allow to test the ui, which can't be done atm since the templates and the logic reside in separate bundles.

@OskarStark, @dbu Do you think this is the thing to do?


Thanks and have a nice day !

greg0ire commented:
> To do this, however I still need an answer (or any kind of input really, even a global IDKWTFYTA will be helpful) to my previous question.

Sorry, but IDKWTFYTA , I don't know that part very well…

dbu commented:
big +1 for sharing this at highest possible level. every persistence likely has some kind of "related model" concept. and the specific admin bundles are free to not use some of the form types if they don't make sense, so i think even if mongo has no use for a many-to-many widget, it does not hurt if its globally available.

re #issuecomment-211441695 also a +1 on simplifying and unifying things and having less special cases. lets try to have as few different cases as possible to cover standard use cases and ideally have a way to customize things in applications for special cases. sonata admin should focus on doing the standard 80% of an admin and not be in the way to do custom things for the other 20%.

Soullivaneuh commented:
@ju1ius I would love to look at your refactoring PR after ending new version management setup for **all packages** of sonata, but not before.

Why? Because your PR is a big change. I'm sure you understand. ;-)

BTW, I did a very quick review for now. It looks indeed nice, but for me, the PR is huge too big.

149 files changed ATM, this is hard to have a good review for humans.

Could this be spitted onto several PR? One by "section"?

Another question: Could this be done on a BC way? If we can do this it would give a very smooth upgrade workflow and make user happy! :-)

Regards.

core23 commented:
> Another question: Could this be done on a BC way? If we can do this it would give a very smooth upgrade workflow and make user happy! :-)

IMO, not possible without ending in huge code waste that nobody will understand. This is a big refactoring that needs to be done with a BC break, regardless of any consequences.

greg0ire commented:
I think @Soullivaneuh actually meant : make one more PR which would be BC and would contain every thing that is BC in this PR **and** adds deprecation warnings for what isn't.

ju1ius commented:
> I would love to look at your refactoring PR after ending new version management setup for all packages of sonata, but not before.

@Soullivaneuh This is totally the right thing to do, since this PR cannot be merged without merging it's counterparts for the 4 persistence bundles.

ju1ius commented:
> IMO, not possible without ending in huge code waste that nobody will understand. This is a big refactoring that needs to be done with a BC break, regardless of any consequences.

Totally agree with @core23.
First [I already tried](https://github.com/sonata-project/SonataAdminBundle/pull/3647#issuecomment-199193155) in this PR, and nobody seemed interested.
Second, even with a JS compatibility layer, the only thing you can be sure of is that «legacy» userland code won't trigger exceptions. If you want a 100% BC compat layer, at some point you'll end up having the initialization code called twice. Footgun. :gun:

ju1ius commented:
Anyway, concerning this PR, after discussing shortly with @dbu on slack, I came to the conclusion that the first thing to do should be to move most - if not all - of the form templating code currently residing in the persistence bundles back to the admin bundle. So I think I'll have to rework this from the ground up.
This could be the plan:

* A PR to admin bundle + one to each persistence bundle, to move the form templates
* Then one PR to:
  1. add the infrastructure to allow css/js modularization (the npm + webpack + babel + postcss build setup)
  2. split the css file into smaller modules (just reorganization, no code changes)
* Then one PR to remove dead css
* Then one with the big js/html/css refactorings
* And maybe a few small others to add the others small features of this PR

Does this sound right ?

core23 commented:
Sounds like a good process to shrink this big PR 👍

Soullivaneuh commented:
@ju1ius I'm not sure about BC possible or not. You spiting plan sounds good but the best is too see your small PRs and have a better visibility.

> add the infrastructure to allow css/js modularization (the npm + webpack + babel + postcss build setup)

Are you talking about making sonata asset package to get this outside the bundles?

SonataCI commented:
Could you please rebase your PR and fix merge conflicts?

SonataCI commented:
Could you please rebase your PR and fix merge conflicts?

SonataCI commented:
Could you please rebase your PR and fix merge conflicts?

ju1ius commented:
I just started splitting this PR into smaller ones, so I'm closing this one.
See #3972
